motleycrew/agents/parent.py

```diff
@@ -97,12 +110,65 @@ def agent(self):
     def is_materialized(self):
         return self._agent is not None
 
+    def _prepare_output_handler(self) -> Optional[MotleyTool]:
+        """
+        Wraps the output handler in one more tool layer,
+        adding the necessary stuff for returning direct output.
+        Expects agent's later invocation through _run_and_catch_output method below.
+        """
+        if not self.output_handler:
+            return None
+
+        def handle_agent_output(*args, **kwargs):
+            assert self.output_handler
+            try:
+                output = self.output_handler._run(*args, **kwargs)
+            except InvalidOutput as exc:
+                return f"{exc.__class__}: {str(exc)}"
+
+            raise DirectOutput(output)
+
+        prepared_output_handler = StructuredTool.from_function(
+            name=self.output_handler.name,
+            description=self.output_handler.description,
+            func=handle_agent_output,
+            args_schema=self.output_handler.args_schema,
+        )
+        return MotleyTool.from_langchain_tool(prepared_output_handler)
+
+    @staticmethod
+    def _run_and_catch_output(
+        action: Callable, action_args: tuple, action_kwargs: Dict[str, Any]
+    ) -> Any:
+        """
+        Catcher for the direct output from the output handler (see _prepare_output_handler).
+
+        Args:
+            action (Callable): the action inside which the output handler is supposed to be called.
+                Usually the invocation method of the underlying agent.
+            action_args (tuple): the args for the action
+            action_kwargs (tuple): the kwargs for the action
+
+        Returns:
+            output (Any): the output, either caught via DirectOutput or returned by the action
+        """
+        assert callable(action)
+
+        try:
+            output = action(*action_args, **action_kwargs)
+        except DirectOutput as output_exc:
+            return output_exc.output
+
+        return output
+
     def materialize(self):
         if self.is_materialized:
             logger.info("Agent is already materialized, skipping materialization")
             return
         assert self.agent_factory, "Cannot materialize agent without a factory provided"
-        self._agent = self.agent_factory(tools=self.tools)
+        self._agent = self.agent_factory(
+            tools=self.tools, output_handler=self._prepare_output_handler()
```

Comment:
Why not supply output handler as just another tool? We want agent factories to be simple to create, not bother with extra args